File: frontend/src/lib/typeChart.ts
// Gen 6+ type effectiveness chart
// effectiveness[attackingType][defendingType] = multiplier
export const TYPE_CHART: Record<string, Record<string, number>> = {
  normal:   { rock: 0.5, ghost: 0, steel: 0.5 },
  fire:     { fire: 0.5, water: 0.5, grass: 2, ice: 2, bug: 2, rock: 0.5, dragon: 0.5, steel: 2 },
  water:    { fire: 2, water: 0.5, grass: 0.5, ground: 2, rock: 2, dragon: 0.5 },
  electric: { water: 2, electric: 0.5, grass: 0.5, ground: 0, flying: 2, dragon: 0.5 },
  grass:    { fire: 0.5, water: 2, grass: 0.5, poison: 0.5, ground: 2, flying: 0.5, bug: 0.5, rock: 2, dragon: 0.5, steel: 0.5 },
  ice:      { fire: 0.5, water: 0.5, grass: 2, ice: 0.5, ground: 2, flying: 2, dragon: 2, steel: 0.5 },
  fighting: { normal: 2, ice: 2, poison: 0.5, flying: 0.5, psychic: 0.5, bug: 0.5, rock: 2, ghost: 0, dark: 2, steel: 2, fairy: 0.5 },
  poison:   { grass: 2, poison: 0.5, ground: 0.5, rock: 0.5, ghost: 0.5, steel: 0, fairy: 2 },
  ground:   { fire: 2, electric: 2, grass: 0.5, poison: 2, flying: 0, bug: 0.5, rock: 2, steel: 2 },
  flying:   { electric: 0.5, grass: 2, fighting: 2, bug: 2, rock: 0.5, steel: 0.5 },
  psychic:  { fighting: 2, poison: 2, psychic: 0.5, dark: 0, steel: 0.5 },
  bug:      { fire: 0.5, grass: 2, fighting: 0.5, flying: 0.5, psychic: 2, ghost: 0.5, dark: 2, steel: 0.5, fairy: 0.5 },
  rock:     { fire: 2, ice: 2, fighting: 0.5, ground: 0.5, flying: 2, bug: 2, steel: 0.5 },
  ghost:    { normal: 0, psychic: 2, ghost: 2, dark: 0.5 },
  dragon:   { dragon: 2, steel: 0.5, fairy: 0 },
  dark:     { fighting: 0.5, psychic: 2, ghost: 2, dark: 0.5, fairy: 0.5 },
  steel:    { fire: 0.5, water: 0.5, electric: 0.5, ice: 2, rock: 2, steel: 0.5, fairy: 2 },
  fairy:    { fire: 0.5, fighting: 2, poison: 0.5, dragon: 2, dark: 2, steel: 0.5 },
};

export const ALL_TYPES = [
  "normal","fire","water","electric","grass","ice","fighting","poison",
  "ground","flying","psychic","bug","rock","ghost","dragon","dark","steel","fairy",
];

export function getEffectiveness(attackType: string, defTypes: string[]): number {
  let mult = 1;
  for (const defType of defTypes) {
    const row = TYPE_CHART[attackType] ?? {};
    mult *= row[defType] ?? 1;
  }
  return mult;
}

export type CoverageResult = {
  type: string;
  multiplier: number;
};

/** For a team's offensive types, find best multiplier against each defending type combo */
export function getTeamOffensiveCoverage(
  teamTypes: string[][]
): Record<string, number> {
  const result: Record<string, number> = {};
  for (const defType of ALL_TYPES) {
    let best = 0;
    for (const pokemonTypes of teamTypes) {
      for (const atkType of pokemonTypes) {
        const eff = getEffectiveness(atkType, [defType]);
        if (eff > best) best = eff;
      }
    }
    result[defType] = best;
  }
  return result;
}

/** For a defending team's type combo, find weaknesses from all attack types */
export function getTeamDefensiveWeaknesses(
  teamDefTypes: string[][]
): Record<string, number[]> {
  // Returns: for each pokemon, map of attacker_type -> multiplier
  return Object.fromEntries(
    teamDefTypes.map((defTypes, i) => {
      const weaknesses: Record<string, number> = {};
      for (const atkType of ALL_TYPES) {
        weaknesses[atkType] = getEffectiveness(atkType, defTypes);
      }
      return [i.toString(), Object.values(weaknesses)];
    })
  );
}

File: backend/app/battle/typechart.py
"""Type effectiveness chart and lookup (Gen VI+, 18 types including Fairy)."""

from __future__ import annotations

TYPE_CHART: dict[str, dict[str, float]] = {
    "normal": {"rock": 0.5, "ghost": 0.0, "steel": 0.5},
    "fire": {
        "fire": 0.5,
        "water": 0.5,
        "rock": 0.5,
        "dragon": 0.5,
        "grass": 2.0,
        "ice": 2.0,
        "bug": 2.0,
        "steel": 2.0,
    },
    "water": {"water": 0.5, "grass": 0.5, "dragon": 0.5, "fire": 2.0, "ground": 2.0, "rock": 2.0},
    "electric": {
        "electric": 0.5,
        "grass": 0.5,
        "dragon": 0.5,
        "ground": 0.0,
        "water": 2.0,
        "flying": 2.0,
    },
    "grass": {
        "fire": 0.5,
        "grass": 0.5,
        "poison": 0.5,
        "flying": 0.5,
        "bug": 0.5,
        "dragon": 0.5,
        "steel": 0.5,
        "water": 2.0,
        "ground": 2.0,
        "rock": 2.0,
    },
    "ice": {
        "fire": 0.5,
        "water": 0.5,
        "ice": 0.5,
        "steel": 0.5,
        "grass": 2.0,
        "ground": 2.0,
        "flying": 2.0,
        "dragon": 2.0,
    },
    "fighting": {
        "poison": 0.5,
        "flying": 0.5,
        "psychic": 0.5,
        "bug": 0.5,
        "fairy": 0.5,
        "ghost": 0.0,
        "normal": 2.0,
        "ice": 2.0,
        "rock": 2.0,
        "dark": 2.0,
        "steel": 2.0,
    },
    "poison": {
        "poison": 0.5,
        "ground": 0.5,
        "rock": 0.5,
        "ghost": 0.5,
        "steel": 0.0,
        "grass": 2.0,
        "fairy": 2.0,
    },
    "ground": {
        "grass": 0.5,
        "bug": 0.5,
        "flying": 0.0,
        "fire": 2.0,
        "electric": 2.0,
        "poison": 2.0,
        "rock": 2.0,
        "steel": 2.0,
    },
    "flying": {
        "electric": 0.5,
        "rock": 0.5,
        "steel": 0.5,
        "grass": 2.0,
        "fighting": 2.0,
        "bug": 2.0,
    },
    "psychic": {"psychic": 0.5, "steel": 0.5, "dark": 0.0, "fighting": 2.0, "poison": 2.0},
    "bug": {
        "fire": 0.5,
        "fighting": 0.5,
        "poison": 0.5,
        "flying": 0.5,
        "ghost": 0.5,
        "steel": 0.5,
        "fairy": 0.5,
        "grass": 2.0,
        "psychic": 2.0,
        "dark": 2.0,
    },
    "rock": {
        "fighting": 0.5,
        "ground": 0.5,
        "steel": 0.5,
        "fire": 2.0,
        "ice": 2.0,
        "flying": 2.0,
        "bug": 2.0,
    },
    "ghost": {"normal": 0.0, "dark": 0.5, "psychic": 2.0, "ghost": 2.0},
    "dragon": {"steel": 0.5, "fairy": 0.0, "dragon": 2.0},
    "dark": {"fighting": 0.5, "dark": 0.5, "fairy": 0.5, "psychic": 2.0, "ghost": 2.0},
    "steel": {
        "fire": 0.5,
        "water": 0.5,
        "electric": 0.5,
        "steel": 0.5,
        "ice": 2.0,
        "rock": 2.0,
        "fairy": 2.0,
    },
    "fairy": {
        "fire": 0.5,
        "poison": 0.5,
        "steel": 0.5,
        "fighting": 2.0,
        "dragon": 2.0,
        "dark": 2.0,
    },
}


def get_type_effectiveness(attacking: str, defending_types: list[str]) -> float:
    """Calculate type effectiveness multiplier for an attack against defending types."""
    chart = TYPE_CHART.get(attacking, {})
    mult = 1.0
    for def_type in defending_types:
        mult *= chart.get(def_type, 1.0)
    return mult
